### PR TITLE
Add simple probabilistic query tracing and logging for streaming searches that time out.

### PR DESCRIPTION
**Pull request by vekterli**
@baldersheim @geirst please review. This is a poor man's version of distributed tracing, meant as a stop-gap for debugging until we have any integrations of OpenTelemetry et al.

This currently only catches the cases where a query has not respected its
expected timeout and therefore has an actual timeout that is at least a
5x multiple of its expected one. Due to the response payload nature of
MessageBus traces it's not given that the trace will contain any
information from the backend nodes. But it is highly likely to contain
useful information from the client-local policy instances, as well as
resending info.

To test this very conservatively, only allows approximately 1 query to be
traced every 2 seconds and only dumps traces to logs once every 10 seconds.

### CODE REVIEW

**Review by geirst — APPROVED** [review]
👍 Some nitpicks to consider.

File: container-search/src/main/java/com/yahoo/vespa/streamingvisitors/TracingOptions.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+import com.yahoo.vespa.streamingvisitors.tracing.MaxSamplesPerPeriod;
+import com.yahoo.vespa.streamingvisitors.tracing.MonotonicNanoClock;
+import com.yahoo.vespa.streamingvisitors.tracing.ProbabilisticSampleRate;
+import com.yahoo.vespa.streamingvisitors.tracing.SamplingStrategy;
+import com.yahoo.vespa.streamingvisitors.tracing.SamplingTraceExporter;
+import com.yahoo.vespa.streamingvisitors.tracing.TraceExporter;
+
+import java.util.concurrent.TimeUnit;
+
+public class TracingOptions {
```
[the remark this answers is not in the document]
Comment (vekterli) [reply container-search/src/main/java/com/yahoo/vespa/streamingvisitors/TracingOptions.java]: Done

File: container-search/src/main/java/com/yahoo/vespa/streamingvisitors/tracing/MonotonicNanoClock.java [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2019 Oath Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.streamingvisitors.tracing;
+
+/*
```
[the remark this answers is not in the document]
Comment (vekterli) [reply container-search/src/main/java/com/yahoo/vespa/streamingvisitors/tracing/MonotonicNanoClock.java]: Fixed

File: container-search/src/main/java/com/yahoo/vespa/streamingvisitors/VdsStreamingSearcher.java [hunk 3]
```diff
@@ -73,26 +76,105 @@ public final void setStorageClusterRouteSpec(String storageClusterRouteSpec) {
...
+            return new Result(query, ErrorMessage.createBackendCommunicationError(
+                    "Failed to parse document selection string: " + e.getMessage() + "'."));
+        } catch (TokenMgrException e) {
+            return new Result(query, ErrorMessage.createBackendCommunicationError(
+                    "Failed to tokenize document selection string: " + e.getMessage() + "'."));
+        } catch (TimeoutException e) {
+            double elapsedMillis = durationInMillisFromNanoTime(timeStartedNanos);
+            if ((effectiveTraceLevel > 0) && timeoutBadEnoughToBeReported(query, elapsedMillis)) {
+                tracingOptions.getTraceExporter().maybeExport(() -> new TraceDescription(visitor.getTrace(),
+                        String.format("Trace of %s which timed out after %.2g ms",
```
[the remark this answers is not in the document]
Comment (vekterli) [reply container-search/src/main/java/com/yahoo/vespa/streamingvisitors/VdsStreamingSearcher.java]: Done

**Review by baldersheim — APPROVED** [review]

**Comment by baldersheim** [comment]
Javadoc not happy
[ERROR] Exit code: 1 - /source/container-search/src/main/java/com/yahoo/vespa/streamingvisitors/TracingOptions.java:38: error: bad use of '>'
[ERROR]      * @param traceTimeoutMultiplierThreshold only export traces if the elapsed time is > query timeout * this value
[ERROR]                                                                                         ^
[ERROR] 
[ERROR] Command line was: /usr/lib/jvm/java/bin/javadoc @options @packages

**Pull request merged**